petab/v2/core.py

```diff
@@ -548,12 +547,20 @@ def from_df(cls, df: pd.DataFrame) -> ExperimentTable:
 
         experiments = []
         for experiment_id, cur_exp_df in df.groupby(C.EXPERIMENT_ID):
-            periods = [
-                ExperimentPeriod(
-                    time=row[C.TIME], condition_id=row[C.CONDITION_ID]
+            periods = []
+            for timepoint in cur_exp_df[C.TIME].unique():
+                condition_ids = [
+                    cid
+                    for cid in cur_exp_df.loc[
+                        cur_exp_df[C.TIME] == timepoint, C.CONDITION_ID
+                    ]
+                    if not pd.isna(cid)
+                ]
```

Comment:
Might need to add a `dropna`
```suggestion
            for timepoint, cur_exp_time_df in cur_exp_df.groupby(C.TIME):
                condition_ids = cur_exp_time_df[C.CONDITION_ID].unique()
```

Reply:
I wanted to avoid the overhead of the additional groupy / creating the additional dataframes. Usually there will only be a single condition ID.

I'd avoid the `unique` otherwise this will hide errors in case of duplicated conditions.

Reply:
> creating the additional dataframes

Ah, you're right, I thought it was some view instead

> I'd avoid the `unique`

Yes, not sure why I did that...

No issue with your code but I find the groupby faster to understand.
Independently of this, again this might be something that tools want in a library function